File: python-sdk/src/forta_agent/finding.py
import json
from .label import Label
from enum import IntEnum
from .utils import assert_enum_value_in_dict, assert_non_empty_string_in_dict


class FindingSeverity(IntEnum):
    Unknown = 0
    Info = 1
    Low = 2
    Medium = 3
    High = 4
    Critical = 5


class FindingType(IntEnum):
    Unknown = 0
    Exploit = 1
    Suspicious = 2
    Degraded = 3
    Info = 4


class Finding:
    def __init__(self, dict):
        assert_non_empty_string_in_dict(dict, 'name')
        assert_non_empty_string_in_dict(dict, 'description')
        assert_non_empty_string_in_dict(dict, 'alert_id')
        assert_enum_value_in_dict(dict, 'severity', FindingSeverity)
        assert_enum_value_in_dict(dict, 'type', FindingType)
        self.name = dict['name']
        self.description = dict['description']
        self.alert_id = dict['alert_id']
        self.protocol = dict.get('protocol', 'ethereum')
        self.severity = dict['severity']
        self.type = dict['type']
        self.metadata = dict.get('metadata')
        self.addresses = dict.get('addresses')
        self.labels = list(map(lambda l: l if isinstance(l, Label) else Label(l), dict.get('labels', [])))

    def toJson(self):
        d = dict(self.__dict__, **{
            'alertId': self.alert_id,
            'labels': list(map(lambda l: l.toDict(), self.labels))
        })
        return json.dumps({k: v for k, v in d.items() if v or k == 'type' or k == 'severity'})

File: sdk/finding.ts
import { EntityType, Label } from "./label"
import { assertIsFromEnum, assertIsNonEmptyString } from "./utils"

export enum FindingSeverity {
  Unknown,
  Info,
  Low,
  Medium,
  High,
  Critical
}

export enum FindingType {
  Unknown,
  Exploit,
  Suspicious,
  Degraded,
  Info,
  Scam
}

type FindingInput = {
  name: string,
  description: string,
  alertId: string,
  protocol?: string,
  severity: FindingSeverity,
  type: FindingType,
  metadata?: { [key: string]: string},
  addresses?: string[],
  labels?: Label[],
  uniqueKey?: string,
}

export class Finding {
  private constructor(
    readonly name: string,
    readonly description: string,
    readonly alertId: string,
    readonly protocol: string,
    readonly severity: FindingSeverity,
    readonly type: FindingType,
    readonly metadata: { [key: string]: string},
    readonly addresses: string[],
    readonly labels: Label[],
    readonly uniqueKey: string,
  ) {}

  toString() {
    return JSON.stringify({
      ...this,
      severity: FindingSeverity[this.severity],
      type: FindingType[this.type],
      labels: this.labels.map(l => Object.assign(l, {
        entityType: EntityType[l.entityType]}))
    }, null, 2)
  }

  static from(findingInput: FindingInput) {
    return this.fromObject(findingInput)
  }

  static fromObject({
    name,
    description,
    alertId,
    protocol = 'ethereum',
    severity,
    type,
    metadata = {},
    addresses = [],
    labels = [],
    uniqueKey = '',
  }: FindingInput) {
    assertIsNonEmptyString(name, 'name')
    assertIsNonEmptyString(description, 'description')
    assertIsNonEmptyString(alertId, 'alertId')
    assertIsNonEmptyString(protocol, 'protocol')
    assertIsFromEnum(severity, FindingSeverity, 'severity')
    assertIsFromEnum(type, FindingType, 'type')
    // TODO assert metadata keys and values are strings

    labels = labels.map(l => l instanceof Label ? l : Label.fromObject(l))
    return new Finding(name, description, alertId, protocol, severity, type, metadata, addresses, labels, uniqueKey)
  }
}